Use RocksDB's snapshots instead of RwLock on database

jakrawcz-rdx opened the pull request:
The first PR for https://radixdlt.atlassian.net/browse/NODE-571.
Addresses point **1.** of DoD: just not via a single hot-swapped snapshot, but by on-demand snapshots.

Main points:
- `StateLock<>` is now replaced by `DbLock<Snapshottable>` (see its extensive rustdocs)
- This lock, apart from "locked" vs "historical" access, can give you a `snapshot()` at any moment:
  - taking a snapshot is apparently so cheap (~microseconds) that it's not worth caching
  - if it turns out otherwise, "add a cache for snapshot-after-last-commit" will be a cool future performance improvement PR
- The "DB enum middleman" (enum_dispatch-ing to rocks only) is gone
- For technical reasons, we need to define our own trait over Rocks (see `ReadableRocks` and `WriteableRocks`)
- Every place that used `StateManagerDatabase` now uses either `StateManagerDatabase<impl ReadableRocks>` or `StateManagerDatabase<impl WriteableRocks>` (depending on its use-case)
  - This gives some extra compile-time safety, so that you don't write to Rocks when you e.g. access it via snapshot.
- **This refactoring PR does _not_ attempt to fix all "sloppy DB locking" problems that we have**; there are cases which (I believe) only work fine because a higher layer synchronizes them (e.g. Java lock) or their nature excludes concurrent processes (e.g. during boot-up). Most notably:
  - execution of genesis and scenarios
  - applying of protocol update

LukasGasior1 commented on core-rust/state-manager/src/store/rocks_db.rs:
```diff
@@ -596,106 +812,39 @@ impl RocksDBStore {
         }
     }
 
     pub fn try_catchup_with_primary(&self) {
```
since we're going the traits / type safety route would it make sense to introduce `SecondaryRocks` and move it there? not that it matters much, but just for consistency

jakrawcz-rdx replied:
good suggestion; I also got rid of the non-sensical `ReadonlyRocks` struct 👍

LukasGasior1 reviewed (approved):
Just a few minor comments.

Looks good! Love the new readable/writeable traits :+1:

jakrawcz-rdx commented on core-rust/node-common/src/locks.rs:
```diff
@@ -246,72 +250,75 @@ impl<T> RwLock<T> {
+    ///
+    /// This method should be used by clients who need to coordinate an exclusive read+write access
+    /// to a known mutable region of the database.
+    // TODO(future enhancement): we really should have a set of `RwLock`s for independent regions?
```
ok: https://radixdlt.atlassian.net/browse/NODE-571

LukasGasior1 commented on core-rust/state-manager/src/store/rocks_db.rs:
```diff
@@ -489,27 +493,203 @@ impl fmt::Display for ExtensionsDataKey {
+/// snapshots.
+///
+/// The library we use (a thin C wrapper, really) does not introduce this trivial and natural trait
+/// itself, while we desperately need it to abstract the DB-reading code from the actual source of
```
Ah gotcha :+1:  I thought it was close to complete

jakrawcz-rdx replied:
it's not rocket science, but yeah, it would need exploring what other methods should belong to such trait (and I know there are a few)

LukasGasior1 commented on core-rust/state-manager/src/state_manager.rs:
```diff
@@ -143,23 +143,17 @@ impl StateManagerConfig {
-        *write_mempool = PriorityMempool::new(MempoolConfig::default(), metrics_registry);
-        drop(write_mempool);
-    }
-
```
_(note from author of this snippet)_

Yes, that was a debug leftover. Thanks :+1:

LukasGasior1 commented on core-rust/core-api-server/src/core_api/handlers/lts/transaction_status.rs:
```diff
@@ -30,9 +30,7 @@ pub(crate) async fn handle_lts_transaction_status(
         pending_transaction_result_cache.peek_all_known_payloads_for_intent(&intent_hash);
     drop(pending_transaction_result_cache);
 
-    // TODO(locks): consider this (and other) usages of DB "read current" lock. *Carefully* migrate
```
Hmh so we're now allowing (locking-wise) concurrent writes to VertexStoreCf (in `save_vertex_store` and `commit`).
While I don't think it can cause issues at the moment given the context in which those methods are used (e.g. we only include vertex store in `commit` if it originates from consensus, and that's the same thread that calls `save_vertex_store`), I'd still suggest to change that to `lock` just in case (e.g. if some assumptions change in the future). In normal operating conditions this lock will always be non-contentious.

jakrawcz-rdx replied:
yeah, let me add the lock, since it indeed writes to the same "db region" as `commit` 👍

jakrawcz-rdx commented:
personally I was waiting here for the performance tests - they are now complete (at https://radixdlt.atlassian.net/browse/NODE-593) and they confirm that there is some improvement in Core API response times (with no degradation to consensus TPS).

@LukasGasior1 are you happy with the code after addressing the comments? plz ✅

LukasGasior1 reviewed (approved):
LGTM
